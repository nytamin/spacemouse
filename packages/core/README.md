# SpaceMouse - Core

This package contains the platform-agnostic parts of the SpaceMouse library.

*You should not be importing this package directly, instead you'll want to use one of the wrapper libraries to provide the appropriate HID bindings for your target platform, see [https://github.com/nytamin/spacemouse](https://github.com/nytamin/spacemouse)*

