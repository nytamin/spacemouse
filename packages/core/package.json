{
  "name": "@spacemouse-lib/core",
  "version": "0.0.1",
  "description": "NPM package to interact with the SpaceMouse devices",
  "main": "dist/index.js",
  "typings": "dist/index.d.ts",
  "license": "MIT",
  "homepage": "https://github.com/nytamin/spacemouse",
  "repository": {
    "type": "git",
    "url": "git+https://github.com/nytamin/spacemouse.git"
  },
  "bugs": {
    "url": "https://github.com/nytamin/spacemouse/issues"
  },
  "author": {
    "name": "Johan Nyman",
    "email": "johan@nytamin.se",
    "url": "https://github.com/nytamin"
  },
  "scripts": {
    "build": "rimraf dist && yarn build:main",
    "build:main": "tsc -p tsconfig.build.json",
    "test": "jest"
  },
  "files": [
    "dist/**"
  ],
  "engines": {
    "node": ">=14"
  },
  "prettier": "@sofie-automation/code-standard-preset/.prettierrc.json",
  "keywords": [
    "spaceMouse",
    "SpaceMouse",
    "hid",
    "usb",
    "hardware",
    "interface",
    "controller"
  ],
  "dependencies": {
    "tslib": "^2.4.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
