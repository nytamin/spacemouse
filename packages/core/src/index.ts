export * from './lib'
export * from './api'
export * from './products'
export * from './genericHIDDevice'
export * from './SpaceMouse'
