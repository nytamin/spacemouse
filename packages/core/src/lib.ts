/*
 * This file contains internal convenience functions
 */

/** Convenience function to force the input to be of a certain type. */
export function literal<T>(o: T): T {
	return o
}

/** Get bit value in integer */
export function getBit(integer: number, bitPosition: number): 0 | 1 {
	return (integer & (1 << bitPosition)) === 0 ? 0 : 1
}
