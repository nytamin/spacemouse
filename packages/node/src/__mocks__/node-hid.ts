import { EventEmitter } from 'events'

export interface Device {
	vendorId: number
	productId: number
	path?: string
	serialNumber?: string
	manufacturer?: string
	product?: string
	release: number
	interface: number
	usagePage?: number
	usage?: number
}

let mockWriteHandler: undefined | ((hid: HID, message: number[]) => void) = undefined
export function setMockWriteHandler(handler: (hid: HID, message: number[]) => void) {
	mockWriteHandler = handler
}

const MockHIDDevices: HID[] = []
export function getMockHIDDevices(): HID[] {
	return MockHIDDevices
}
export function resetMockHIDDevices(): HID[] {
	return MockHIDDevices
}

// export class HID extends EventEmitter {
export class HID extends EventEmitter {
	private mockWriteHandler

	constructor(_path: string) {
		super()
		MockHIDDevices.push(this)
		this.mockWriteHandler = mockWriteHandler
	}
	// constructor(vid: number, pid: number);
	close(): void {
		// void
	}
	pause(): void {
		// void
	}
	read(_callback: (err: any, data: number[]) => void): void {
		// void
	}
	readSync(): number[] {
		return []
	}
	readTimeout(_timeOut: number): number[] {
		return []
	}
	sendFeatureReport(_data: number[]): number {
		return 0
	}
	getFeatureReport(_reportIdd: number, _reportLength: number): number[] {
		return []
	}
	resume(): void {
		// void
	}
	write(message: number[]): number {
		this.mockWriteHandler?.(this, message)
		return 0
	}
	setNonBlocking(_noBlock: boolean): void {
		// void
	}
}
export function devices(): Device[] {
	return []
}
export function setDriverType(_type: 'hidraw' | 'libusb'): void {
	// void
}
