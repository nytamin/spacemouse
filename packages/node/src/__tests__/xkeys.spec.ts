import * as HID from 'node-hid'
import * as HIDMock from '../__mocks__/node-hid'
import { PRODUCTS, Rotation, setupSpaceMouse } from '../'
import { getSentData, handleSpaceMouseMessages, resetSentData } from './lib'
import { getMockHIDDevices } from '../__mocks__/node-hid'
import { Translation } from '@spacemouse-lib/core'

describe('Unit tests', () => {
	test('SpaceMouse methods', async () => {
		// const device = new SpaceMouse()

		const hidDevice = {
			vendorId: PRODUCTS['SpaceNavigator'].vendorId,
			productId: PRODUCTS['SpaceNavigator'].productId,
			interface: 0,
			path: 'mockPath',
		} as HID.Device

		HIDMock.setMockWriteHandler(handleSpaceMouseMessages)

		const myspaceMouse = await setupSpaceMouse(hidDevice)

		const onError = jest.fn(console.log)

		myspaceMouse.on('error', onError)

		let rotate: Rotation = {
			pitch: 0,
			roll: 0,
			yaw: 0,
		}
		myspaceMouse.on('rotate', (r) => {
			rotate = r
		})
		let translate: Translation = {
			x: 0,
			y: 0,
			z: 0,
		}
		myspaceMouse.on('translate', (t) => {
			translate = t
		})

		resetSentData()

		resetSentData()
		expect(myspaceMouse.info).toMatchSnapshot()
		resetSentData()
		myspaceMouse.getButtons()
		expect(getSentData()).toMatchSnapshot()
		resetSentData()

		const mockHIDs = getMockHIDDevices()
		expect(mockHIDs).toHaveLength(1)
		const mockHID = mockHIDs[0]

		// Rotation:
		mockHID.emit('data', Buffer.from([0x02, 0x32, 0x00, 0xff, 0xff, 0x8b, 0x00]))
		expect(onError).toHaveBeenCalledTimes(0)
		expect(rotate).toEqual({
			pitch: 50,
			roll: -1,
			yaw: 139,
		})
		// Translation:
		mockHID.emit('data', Buffer.from([0x01, 0x1b, 0x00, 0x15, 0x00, 0x5e, 0x01]))
		expect(onError).toHaveBeenCalledTimes(0)
		expect(translate).toEqual({
			x: 27,
			y: 21,
			z: 350,
		})
	})
})
