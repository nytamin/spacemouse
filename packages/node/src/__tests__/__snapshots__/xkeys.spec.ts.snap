// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests SpaceMouse methods 1`] = `
Object {
  "interface": 50726,
  "name": "SpaceNavigator",
  "productId": 1133,
  "vendorId": 1133,
}
`;

exports[`Unit tests SpaceMouse methods 2`] = `Array []`;
